Allow default_render to take a block to customize behavior when there's no template

davetron5000 opened the pull request:
In 0de4a23 the behavior when there is a missing template was changed to
not raise an error, but instead head :no_content.  This is a breaking
change and some gems rely on this happening.

This adds `raise_on_missing_template` to controllers that, when set to
`true`, enables the old behavior

<hr>
# Previous description for some context

Doing this to have a conversation, so not necessarily proposing this exact patch, but I need help.
## Problem

The patch being reverted here changes `ActionController::Metal::ImplicitRender#default_render` in a way that isn't backwards compatible.  _Before_ this patch, code like this:

``` ruby
# note use of responders gem in here—could be an issue
def create
  user = User.create(params)
  if user.valid?
    redirect_to users_path
  else
    respond_with user # intends to render the `new` template
  end
end
```

Would render the `new` template.  _After_ the patch being reverted here, it does a `HEAD` with no content and error message that the `create` template cannot be found.

So, before, the code was not ever intending to render a `create` template, but the patch being reverted here checks.
## Solution

It seems like the check for the template existing isn't right.  Seems like it should test if something valid would be rendered, and this:

``` ruby
if template_exists?(action_name.to_s, _prefixes, variants: request.variant)
```

doesn't check that.
## Where I need help

What I can't figure out is how `render(*args)` ends up rendering the `new` template, so I can't figure out how to make the `if` statement more consistent with the behavior of `render`.

It could certainly be an issue in the responders gem, but a glance at the definition of `respond_with` ( https://github.com/plataformatec/responders/blob/70bea715bb87c9e7bd402ef88406a72ca327bb12/lib/action_controller/respond_with.rb#L190 ) doesn't lead me to any insights.  In my case https://github.com/plataformatec/responders/blob/70bea715bb87c9e7bd402ef88406a72ca327bb12/lib/action_controller/respond_with.rb#L205 is calling `self.class.responder` which is the default.

Can anyone point me to the right direction?

kaspth commented:
The responders gem is relying on `default_render` (which ends up calling `default_render` on the controller) to [raise if there's a missing template](https://github.com/plataformatec/responders/blob/master/lib/action_controller/responder.rb#L171), which we no longer do.

Maybe we could have `default_render` accept a block, to make it easier for the responders gem to hook in. Something like:

```ruby
def default_render(*args)
  if template_exists?(action_name.to_s, _prefixes, variants: request.variant)
    render(*args)
  else
    block_given? ? yield : default_missing_template_action
  end
end

def default_missing_template_action
  logger.info "No template found for #{self.class.name}\##{action_name}, rendering head :no_content"
  head :no_content
end

# In responders
def to_html
  default_render do
    navigation_behavior
  end
end

def navigation_behavior
  if get?
    controller.default_missing_template_action
  # Other branches...
  end
end
```

cc @josevalim @rafaelfranca.

davetron5000 commented:
I think it's more that the responders gem was relying on `render(*args)` sending the `new` template.  Somewhere inside `render` this is what happens, even now.  If I remove the `if template_exists?` check from Rails, everything works as it did before.  So, in this case, no exception would've been thrown.

In other words, from the point of view of whatever is going on in `render`, there is no missing template.  Therefore the `if template_exists?` check seems too strict (or at least inconsistent with whatever's going on in `render`).

What I couldn't figure out is _where_ Rails decides to use the `new` template in this case.  I was able to find the `render` method, but got really lost in there and hit a dead end.

kaspth commented:
You're looking in the wrong place, Rails doesn't render a `new` template here.

Rails would implicitly have rendered the `create.html.erb` (if it existed) and not `new.html.erb`, that's all the responders gem: https://github.com/plataformatec/responders/blob/master/lib/action_controller/responder.rb#L201

davetron5000 commented:
AH!  I think that's what I was looking for.

Let me just summarize to make sure I'm understanding:

1. My code does `respond_with some_record_that_is_not_valid`
2. responders gem calls `default_render`
3. Rails (4.2) throws `ActionView::MissingTemplate`
4. responders gem catches that
5. responders gem then renders the `new` template

Given that, it seems like the patch I'm reverting here is a breaking change to the rendering code.  I read through the PR that introduced it, and admit to not understanding what problem it solves.

That said, it seems like responders cannot easily continue to work without ability to detect a missing template.

Is it something where we can do:

```ruby
def default_render(*args)
  if template_exists?(action_name.to_s, _prefixes, variants: request.variant)
    render(*args)
  else
    logger.info "No template found for #{self.class.name}\##{action_name}, rendering head :no_content" if logger
    if block_given?
      yield(*args)
    else
      head :no_content
    end
  end
end
```

Or is `default_render` something we should consider a Rails internal and thus plugins/gems should not mess with it?

kaspth commented:
Yes, that's how it works.

`default_render` is public API and it not raising `ActionView::MissingTemplate` is a breaking change. If you look at my first comment I have already proposed a solution, but we're on the same page.

Another solution would be to add a setting to controllers, along the lines of:

```ruby
mattr_accessor :raise_on_missing_template
self.raise_on_missing_template = false # Rails default

# In responders' #respond_to
controller.raise_on_missing_template = true
```

Now that we're on the same page let's shelve the discussion until some of the others have chimed in :smile:

davetron5000 commented:
:+1:

rafaelfranca commented:
@kaspth I like the block version. @davetron5000 could you change your patch to that so we can upgrade responders?

davetron5000 committed:
Override default_render's behavior with a block

In 0de4a23 the behavior when there is a missing template was changed to
not raise an error, but instead head :no_content.  This is a breaking
change and some gems rely on this happening.

To allow gems and other code to work around this, allow
`default_render` to take a block which, if provided, will
execute the contents of that block instead of doing the `head :no_content`.

davetron5000 commented:
I couldn't figure out how to test `default_render` so that I could add the block syntax, so I went with the `mattr_accessor` method.

I'm thinking this might be better.  If we went with the block syntax, we'd need to change responders to use it, and then everyone that used responders gets the raise-on-missing-template error, which may not be desired.

This way, controllers that *want* the old behavior have to opt in.

What @rafaelfranca can help me with is if this thinking is correct, or if responders gem is basically broken without the raise-on-missing-template functionality?

kaspth commented:
@davetron5000 find the tests for `implicit_render.rb`, create a test controller with this action:

```ruby
def show
  default_render do
     render text: 'default_render yielded'
end
```

Then you need a test action which does a `get :show` and calls `assert_equal 'default_render yielded', response.body`. Let me know if this helps you out :smile:

spastorino commented:
head :no_content call shouldn't be placed in this module. I think it would be better to put this call somewhere so controllers not adding ImplicitRender like AC::API can take advantage of the no content response.

kaspth commented:
@spastorino good suggestion. Perhaps we could add a `default_render_fallback` method somewhere? Though I don't know where that would be exactly. Another module API could include?

davetron5000 commented:
Does moving the `head :no_content` call affect the problem identified here?  If so, tell me more about what you are thinking :)

Otherwise, I'm going to focus on this particular issue and leave the disposition of `head :no_content` to another PR

spastorino commented:
@davetron5000 :+1: sounds good, I was just mentioning another issue related to `head :no_content`

kaspth commented:
@davetron5000 we'd really like to use the block version instead of another config. Are you still up for making that happen? :smile:

davetron5000 commented:
@kaspth @spastorino Thanks for the help in figuring this out.

This PR now has an implementation of the block-based way of doing things.  I kept it very minimal to see if this is what you all were thinking, but if it is, I think a few more things need to be done (which I can do):

- [ ] make a new test file, I think—this feature doesn't feel like it should be tested in `respond_to_test.rb` to me
- [ ] more explicitly coverage of the default path
- [ ] documentation on what's going on

(Sorry for the delays—was on vacation!)

kaspth commented on actionpack/test/controller/mime/respond_to_test.rb:
```diff
@@ -793,3 +793,24 @@ def test_variant_negotiation_without_block
+  end
+end
+
+class RespondToWithBlockOnDefaultrenderControllerTest < ActionController::TestCase
```
Render with capital R here and in the controller name.

kaspth commented on actionpack/lib/action_controller/metal/implicit_render.rb:
```diff
@@ -10,8 +10,12 @@ def default_render(*args)
+        else
+          logger.info "No template found for #{self.class.name}\##{action_name}, rendering head :no_content" if logger
+          head :no_content
+        end
```
I still think we should do it this way:

```ruby
def default_render(*args)
  if template_exists?(action_name.to_s, _prefixes, variants: request.variant)
    render(*args)
  else
    block_given? ? yield : default_missing_template_action
  end
end

def default_missing_template_action
  logger.info "No template found for #{self.class.name}\##{action_name}, rendering head :no_content"
  head :no_content
end

# In responders
def to_html
  default_render do
    navigation_behavior
  end
end

def navigation_behavior
  if get?
    controller.default_missing_template_action
  # Other branches...
  end
end
```

This way responders doesn't have to reimplement our convention on what happens when a template is missing. Do you agree with this, @rafaelfranca?

kaspth commented:
I think it's okay to test it in `respond_to_test.rb`.

I'm not where you're going with the last two points can you elaborate?

davetron5000 commented:
@kaspth great feedback, thanks.  Check now.

Last two points:

* if this code went to another test file, I kinda feel like it would need more coverage of `default_render`, but if you are cool keeping it where it is, I don't see the need for additional tests, as `default_render` is covered well enough.
* documentation - I made a pass at it, basically wanted to document what passing a block does, as it's likely not obvious w/out reading the code.  Would love to know if what I wrote makes sense

kaspth commented on actionpack/lib/action_controller/metal/implicit_render.rb:
```diff
@@ -6,15 +6,34 @@ def send_action(method, *args)
+        if block_given?
+          yield(*args)
+        else
+          default_missing_template_action(*args)
```
I'm not sure we should pass the args here (I know `default_render` does it, but still). What do you think, @rafaelfranca?

kaspth replied:
I'm trending towards more :-1: about passing the args through here. If people want to know what template was attempted to be rendered they can use `action_name`, and there's a corresponding API for variants.

@davetron5000 if you want to pass the args through you have to find a use case we can evaluate. Otherwise it's more baggage than we need and should be cut :smile:

kaspth commented:
Then let's keep it in the same file, so we don't need to cover something that's already covered :smile:

I'll look at the documentation now.

kaspth commented on actionpack/lib/action_controller/metal/implicit_render.rb:
```diff
@@ -6,15 +6,34 @@ def send_action(method, *args)
+    # If there is no appropriate template, and no block is given, defers to
+    # +default_missing_template_action+.  Otherwise, executes the block.
+    #
+    # +args+:: passed through to render
```
Here's my suggestion:

```ruby
# Renders the template corresponding to the controller action if it exists.
# The action name, format and variant are taken into account.
# This way a new action with an html format and a phone variant would attempt
# to render a new.html+phone.erb template.
#
# If no template is found +default_missing_template_action+ is called. Alternatively,
# +default_render+ takes a block which can override the +default_missing_template_action+.
#
#   default_render do
#     head 404 # No template was found
#   end
```

kaspth commented on actionpack/lib/action_controller/metal/implicit_render.rb:
```diff
@@ -6,15 +6,34 @@ def send_action(method, *args)
+    # When there is no template for a given action, returns an empty HEAD.  To override,
+    # pass a block to +default_render+
+    #
+    # +args+:: passed through from default_render
```
```ruby
# Called when +default_render+ can't find a template to render.
# Renders `head :no_content`.
```

davetron5000 commented on actionpack/lib/action_controller/metal/implicit_render.rb:
```diff
@@ -6,15 +6,40 @@ def send_action(method, *args)
+    #
+    #   default_render do
+    #     head 404 # No template was found
+    #   end
```
@kaspth thanks for the doc suggesting.  Took your version and tweaked slightly.

davetron5000 commented on actionpack/lib/action_controller/metal/implicit_render.rb:
```diff
@@ -6,15 +6,40 @@ def send_action(method, *args)
       end
     end
 
+    # Called when +default_render+ cannot find a template to render.
```
@kaspth took your docs and slightly tweaked.

kaspth commented on actionpack/lib/action_controller/metal/implicit_render.rb:
```diff
@@ -6,15 +6,40 @@ def send_action(method, *args)
+    #     head 404 # No template was found
+    #   end
+    #
+    # +args+:: passed through to render
```
That double colon isn't standard rdoc iirc. I also don't think it communicates very much, which is why I left it out originally.

kaspth commented:
Once my last comments have been addressed this is ready :+1:

davetron5000 commented:
@kaspth removed `args` from that method (didn't have a specific case in mind).  Also removed the rdoc for the args in `default_render`.  FWIW, I think the `::` syntax *does* work to generate a definition list, but since the docs on that argument were not that helpful, I agree it doesn't need to be there.

If this is good, should I squash this and rebase against master?

kaspth commented:
Yes, please. Ping someone who is allowed to merge code, when you have done so :smile:

davetron5000 commented:
ok, squashed and all that.  This is my first PR to Rails, so what's the best way to get it on someone's radar to merge?

davetron5000 commented on actionpack/lib/action_controller/metal/implicit_render.rb:
```diff
@@ -3,12 +3,27 @@ module ImplicitRender
+          yield(*args)
+        else
+          logger.info "No template found for #{self.class.name}\##{action_name}, rendering head :no_content" if logger
+          super
```
Changes on master moved code out of here, so this needs to call `super`, and thus I don't think we can have a `default_missing_template_action` here.

kaspth replied:
Ah, I imagine that's part of the API stuff. c/d @spastorino?

kaspth replied:
Yup, it is. I guess it's fine, though it worries me that this comment says `head :no_content` and super could do something different.

spastorino replied:
@kaspth c

spastorino replied:
We could move the `rendering head :no_content` log to the super class

kaspth replied:
Yeah, I think I'd be more comfortable with that. @davetron5000 can you make that happen?

kaspth commented:
@davetron5000 Congrats and well done!

Just ping someone. I'll do it this time. @rafaelfranca @spastorino @matthewd I think this is ready to merge.

matthewd commented:
@kaspth people should absolutely not just start throwing names into comments

kaspth commented:
Sorry, you're right, Matthew. That gets unwieldy quickly. @davetron5000 please disregard what I said before.

eileencodes commented:
To clarify since @davetron5000 asked - all the members with push access get notifications and emails of all comments and opened pr/issues. No need to ping, someone will merge when they have time. :smile:

davetron5000 commented on actionpack/lib/action_controller/metal/implicit_render.rb:
```diff
@@ -3,12 +3,27 @@ module ImplicitRender
+        if block_given?
+          yield(*args)
+        else
+          logger.info "No template found for #{self.class.name}\##{action_name}" if logger
```
Moving the comment as it was to `BasicImplicitRender` was odd, because that class doesn't really have the notion of a template existing or not, so the comment was out of place.

So, I just removed the "rendering head :no_content" from the message here.  If I understand the reason for this log message to exist, it's to help someone who might be surprised by this behavior, so I think this still achieves that.  The programmer will see that head :no_content was rendered.

kaspth replied:
I think the message sounds like an error if it doesn't say that it's rendering `head :no_content`. I think it would be okay to just leave it as it was.

davetron5000 replied:
no problem.  Will change back.

davetron5000 commented:
@eileencodes thanks for the explanation.  I will not ping anyone else :)

kaspth commented:
I think this is ready :+1:

kaspth commented:
I'll close and reopen for Travis to run, if the build passes I'll merge :+1:

kaspth commented:
I have no idea why Travis says the MySQL client is outdated. Maybe a rebase will help?

rafaelfranca commented:
@kaspth :shipit:, AR failures should not be caused by this PR.

kaspth commented:
@rafaelfranca Alright, :see_no_evil:

davetron5000 commented:
Thanks everyone!

kaspth commented:
@davetron5000 Thanks for bringing this to our attention. Will you be following this up on the responders gem? :smile:

davetron5000 commented:
Yup
